prov/gni/test/cntr.c

```diff
@@ -797,3 +797,189 @@ Test(cntr_mt, write_wait)
+	cr_assert_eq(cntr_val, 0, "write error counter is incorrect.");
+
+	// Create & Spawn threads
+	for (i = 0; i < nthreads; i++) {
+		cr_assert(!pthread_create(threads + i, NULL, do_add_cntr_mt,
+					  (void *) &iters));
+	}
+
+	// Wait until all threads are done
+	for (i = 0; i < nthreads; i++) {
+		cr_assert(!pthread_join(threads[i], &pt_ret));
+		cr_assert(!pt_ret);
+	}
+
+	cntr_val = fi_cntr_read(ep_write_cntrs[0]);
+	cr_assert_eq(cntr_val, (iters * nthreads), "write error counter "
+		"is incorrect.");
+}
+
+Test(cntr_mt, set_add_read_err_cntr)
+{
+	int iters = 128, nthreads = 4, i, ret;
+	uint64_t err_cntr_val;
+	pthread_t threads[nthreads];
+	void *pt_ret;
+
+	ret = fi_cntr_seterr(ep_write_cntrs[0], 0);
+	cr_assert(ret == FI_SUCCESS);
+
+	err_cntr_val = fi_cntr_readerr(ep_write_cntrs[0]);
+	cr_assert_eq(err_cntr_val, 0, "write error counter is incorrect.");
+
+	// Create & Spawn threads
+	for (i = 0; i < nthreads; i++) {
+		cr_assert(!pthread_create(threads + i, NULL, do_add_err_cntr_mt,
+					  (void *) &iters));
+	}
+
+	// Wait until all threads are done
+	for (i = 0; i < nthreads; i++) {
+		cr_assert(!pthread_join(threads[i], &pt_ret));
+		cr_assert(!pt_ret);
+	}
+
+	err_cntr_val = fi_cntr_readerr(ep_write_cntrs[0]);
+	cr_assert_eq(err_cntr_val, (iters * nthreads), "write error counter "
+		"is incorrect.");
+}
+
+//#pragma GCC diagnostic push
+//#pragma diagnostic ignored "-Wpointer-to-int-cast"	
+static void *do_thread_adderr_wait(void *data)
+{
+	int i, ret;
+	i = *((int *) data);
+
+	dbg_printf("%d: waiting\n", i);
+	ret = fi_cntr_wait(ep_write_cntrs[i], ~0, -1);
+	cr_assert(ret != FI_SUCCESS, "Bad return value from fi_cntr_wait");
+
+	dbg_printf("%d: done\n", i);
+	return NULL;
+}
+//#pragma GCC diagnostic pop
+
+//#pragma GCC diagnostic push
+//#pragma diagnostic ignored "-Wint-to-pointer-cast"
+Test(cntr_mt, adderr_wait)
+{
+	int i, ret;
+	pthread_t threads[NUM_EPS];
+	void *pt_ret;
+
+	// Each thread waits for an err cntr change on the i'th ep_write_cntr
+	dbg_printf("creating threads\n");
+	for (i = 0; i < NUM_EPS; i++) {
+		cr_assert(!pthread_create(&threads[i], NULL,
+					  do_thread_adderr_wait, (void *) &i));
+	}
+	
+	dbg_printf("Adding errors\n");
+	for (i = 0; i < NUM_EPS; i++) {
+		ret = fi_cntr_adderr(ep_write_cntrs[i], 1);
+		cr_assert(ret == FI_SUCCESS, "Bad return value from "
+			"fi_cntr_adderr");
+	}
+
+	for (i = 0; i < NUM_EPS; i++) {
+		cr_assert(!pthread_join(threads[i], &pt_ret));
+		cr_assert(!pt_ret);
+	}
+
+	dbg_printf("done\n");
+}
+//#pragma GCC diagnostic pop
+
+Test(cntr, adderr_wait)
+{
+	int ret;
+
+	ret = fi_cntr_adderr(write_cntr, 1);
+	cr_assert(ret == FI_SUCCESS, "Bad return value from fi_cntr_adderr");
+
+	ret = fi_cntr_wait(write_cntr, ~0, -1);
+	cr_assert(ret != FI_SUCCESS, "Bad return value from fi_cntr_wait");
+}
+
+Test(cntr, set_add_read_cntr)
+{
+        int iters = 128, ret, i = 0;
+        uint64_t cntr_val, prev_cntr_val;
+
+        ret = fi_cntr_set(write_cntr, 0);
```

Comment:
Setting the counter to zero and verifying is fine but it doesn't really verify that you wrote something since the default could be zero. I would like you to add another fi_cntr_set to a value that is different from the current value so we verify a set actually happened.